lib/events/dynamoevents/dynamoevents.go

```diff
@@ -625,6 +767,229 @@ func (l *Log) getTableStatus(tableName string) (tableStatus, error) {
 	return tableStatusOK, nil
 }
 
+// indexExists checks if a given index exists on a given table and that it is active.
+func (l *Log) indexExists(tableName, indexName string) (bool, error) {
+	tableDescription, err := l.svc.DescribeTable(&dynamodb.DescribeTableInput{
+		TableName: aws.String(tableName),
+	})
+	if err != nil {
+		return false, trace.Wrap(convertError(err))
+	}
+
+	for _, gsi := range tableDescription.Table.GlobalSecondaryIndexes {
+		if *gsi.IndexName == indexName && *gsi.IndexStatus == dynamodb.IndexStatusActive {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// createV2GSI creates the new global secondary Index and updates
+// the schema to add a string key `date`.
+//
+// This does not remove the old global secondary index.
+// This must be done at a later point in time when all events have been migrated as per RFD 24.
+//
+// Invariants:
+// - The new global secondary index must not exist.
+// - This function must not be called concurrently with itself.
+// - This function must be called before the
+//   backend is considered initialized and the main Teleport process is started.
+func (l *Log) createV2GSI() error {
+	provisionedThroughput := dynamodb.ProvisionedThroughput{
+		ReadCapacityUnits:  aws.Int64(l.ReadCapacityUnits),
+		WriteCapacityUnits: aws.Int64(l.WriteCapacityUnits),
+	}
+
+	// This defines the update event we send to DynamoDB.
+	// This update sends an updated schema and an child event
+	// to create the new global secondary index.
+	c := dynamodb.UpdateTableInput{
+		TableName:            aws.String(l.Tablename),
+		AttributeDefinitions: tableSchema,
+		GlobalSecondaryIndexUpdates: []*dynamodb.GlobalSecondaryIndexUpdate{
+			{
+				Create: &dynamodb.CreateGlobalSecondaryIndexAction{
+					IndexName: aws.String(indexTimeSearchV2),
+					KeySchema: []*dynamodb.KeySchemaElement{
+						{
+							// Partition by date instead of namespace.
+							AttributeName: aws.String(keyDate),
+							KeyType:       aws.String("HASH"),
+						},
+						{
+							AttributeName: aws.String(keyCreatedAt),
+							KeyType:       aws.String("RANGE"),
+						},
+					},
+					Projection: &dynamodb.Projection{
+						ProjectionType: aws.String("ALL"),
+					},
+					ProvisionedThroughput: &provisionedThroughput,
+				},
+			},
+		},
+	}
+
+	if _, err := l.svc.UpdateTable(&c); err != nil {
+		return trace.Wrap(convertError(err))
+	}
+
+	// If we hit this time, we give up waiting.
+	waitStart := time.Now()
+	endWait := waitStart.Add(time.Hour * 24)
+
+	// Wait until the index is created and active.
+	for time.Now().Before(endWait) {
+		indexExists, err := l.indexExists(l.Tablename, indexTimeSearchV2)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if indexExists {
+			log.Info("DynamoDB index created")
+			break
+		}
+
+		time.Sleep(time.Second * 5)
+		elapsed := time.Since(waitStart).Seconds()
+		log.Infof("Creating new DynamoDB index, %f seconds elapsed...", elapsed)
+	}
```

Comment:
24 hours is definitely too long to block (for context, table creation blocks for up to ~8 minutes).  Also, this wait needs to be cancellable by the `context.Context` passed into `New(...)`.

Reply:
Solved now by
- Reducing timeout to 10 minutes
- Marking the index as existing even in an "UPDATE" state which is safe to do, just didn't consider it previously